### PR TITLE
Book details

### PR DESCRIPTION
**Pull request by IgnacioAbadie**
## Summary ##

Added book details view and comments table

## Screenshots ##

ios

## Cards ##
Details -> https://trello.com/c/TTYUOLpw/17-n-61-detalle-del-libro
Comments -> https://trello.com/c/P1sv8ybU/18-n-62-tabla-de-comentarios

### CODE REVIEW

File: WBooks/Extensions/UIAlertControllerExtension.swift [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+//  WBooks
+//
+//  Created by Nacho 2 on 30/04/2020.
+//  Copyright © 2020 Wolox. All rights reserved.
+//
+
+import UIKit
+
+extension UIAlertController {
+    func createErrorAlert(title: String = "", message: String = "", action: @escaping (UIAlertAction) -> Void = {_ in }) -> UIAlertController {
```
Comment (carcos-wolox) [inline WBooks/Extensions/UIAlertControllerExtension.swift]: default values could be nil

File: WBooks/Extensions/UIButtonExtension.swift [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        gradientLayer.endPoint = CGPoint(x: 1, y: 0)
+        gradientLayer.frame = self.bounds
+        gradientLayer.cornerRadius = self.frame.height/2
+        gradientLayer.masksToBounds = false
+
+        self.layer.insertSublayer(gradientLayer, at: 0)
+        self.contentVerticalAlignment = .center
+        self.setTitleColor(UIColor.white, for: .normal)
+        self.titleLabel?.font = UIFont.boldSystemFont(ofSize: 17.0)
+        self.titleLabel?.textColor = UIColor.white
```
Comment (carcos-wolox) [inline WBooks/Extensions/UIButtonExtension.swift]: is it necessary to use both setTitleColor and titleLabel.textColor??
Also.. nice that you made it by yourself, but there is a card for this gradient and you could use the one we have in wolmo core

File: WBooks/Extensions/UITableViewExtension.swift [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+//  Created by Nacho 2 on 07/05/2020.
+//  Copyright © 2020 Wolox. All rights reserved.
+//
+
+import UIKit
+
+extension UITableView {
+
+    func setEmptyMessage(_ message: String) {
+        let messageLabel = UILabel(frame: CGRect(x: 0, y: 0, width: self.bounds.size.width, height: self.bounds.size.height))
```
Comment (carcos-wolox) [inline WBooks/Extensions/UITableViewExtension.swift]: unnecessary self

File: WBooks/Extensions/UITableViewExtension.swift [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        messageLabel.text = message
+        messageLabel.textColor = .black
+        messageLabel.numberOfLines = 0
+        messageLabel.textAlignment = .center
+
+        // messageLabel.font = UIFont(name: "TrebuchetMS", size: 15)
```
Comment (carcos-wolox) [inline WBooks/Extensions/UITableViewExtension.swift]: 👀

File: WBooks/Extensions/UITableViewExtension.swift [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+//
+//  Created by Nacho 2 on 07/05/2020.
+//  Copyright © 2020 Wolox. All rights reserved.
+//
+
+import UIKit
+
+extension UITableView {
+
+    func setEmptyMessage(_ message: String) {
```
Comment (carcos-wolox) [inline WBooks/Extensions/UITableViewExtension.swift:13]: Nice! I have always used a different cell. 👍

File: WBooks/Localizable/es.lproj/Localizable.strings [hunk 6]
```diff
@@ -23,6 +24,15 @@
 "RENTALS_TAB_TITLE" = "Alquileres";
 "SETTINGS_TAB_TITLE" = "Configuración";
 
+// MARK: Book Details View
+
+"RENT" = "ALQUILAR";
+"TO_WISH_LIST" = "AGREGAR A DESEADOS";
+"RENT_UNAVAILABLE" = "Este libro no está disponible por el momento. Por favor, vuelve a intentar mas tarde.";
+"RENT_ERROR" = "Hubo un error. Vuelve  aintentarlo mas tarde.";
```
Comment (carcos-wolox) [inline WBooks/Localizable/es.lproj/Localizable.strings]: Vuelve a intentarlo más

File: WBooks/Localizable/es.lproj/Localizable.strings [hunk 7]
```diff
@@ -23,6 +24,15 @@
 "RENTALS_TAB_TITLE" = "Alquileres";
 "SETTINGS_TAB_TITLE" = "Configuración";
 
+// MARK: Book Details View
+
+"RENT" = "ALQUILAR";
+"TO_WISH_LIST" = "AGREGAR A DESEADOS";
+"RENT_UNAVAILABLE" = "Este libro no está disponible por el momento. Por favor, vuelve a intentar mas tarde.";
```
Comment (carcos-wolox) [inline WBooks/Localizable/es.lproj/Localizable.strings]: más

File: WBooks/Extensions/UIButtonExtension.swift (hunk 2 above)
Comment (IgnacioAbadie) [reply WBooks/Extensions/UIButtonExtension.swift]: nop. Its not needed. I'll remove it.
😱 My gradient is better 🙄

File: WBooks/Extensions/UITableViewExtension.swift (hunk 4 above)
Comment (IgnacioAbadie) [reply WBooks/Extensions/UITableViewExtension.swift]: jeje

File: WBooks/Extensions/UIAlertControllerExtension.swift (hunk 1 above)
Comment (IgnacioAbadie) [reply WBooks/Extensions/UIAlertControllerExtension.swift]: i leave message  required without default and nil for the rest

File: WBooks/Extensions/UITableViewExtension.swift (hunk 5 above)
Comment (IgnacioAbadie) [reply WBooks/Extensions/UITableViewExtension.swift:13]: 💪

File: WBooks/Models/Comment.swift [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    case content
+}
+
+struct Comment: Codable {
+    // MARK: - Properties
+    let id: Int
+    let username: String
+    let image: String?
+    let content: String
+    // MARK: - Codable
```
Comment (carcos-wolox) [inline WBooks/Models/Comment.swift:27]: add line 🙏

File: WBooks/Repositories/BookRepository.swift [hunk 9]
```diff
@@ -28,8 +30,49 @@ class BookRepository {
...
+    
+    public func rentBook(bookId: Int, onSuccess: @escaping () -> Void, onError: @escaping (Error) -> Void) {
+        let endpoint = "\(baseUrl)/users/5/rents"
+        let dateFormatter = DateFormatter()
+        dateFormatter.dateFormat = "yyyy-MM-dd"
+        let params: [String: Any] = [
+            "userID": 5,
+            "bookID": bookId,
+            "from": dateFormatter.string(from: Date()),
+            "to": dateFormatter.string(from: Calendar.current.date(byAdding: .day, value: 1, to: Date()) ?? Date())
```
Comment (carcos-wolox) [inline WBooks/Repositories/BookRepository.swift]: This is business logic, so it must be in the viewModel and get them here as params, you could have a model if you want

File: WBooks/Repositories/BookRepository.swift [hunk 10]
```diff
@@ -28,8 +30,49 @@ class BookRepository {
                 }
             }
     }
+    
+    public func rentBook(bookId: Int, onSuccess: @escaping () -> Void, onError: @escaping (Error) -> Void) {
```
Comment (carcos-wolox) [inline WBooks/Repositories/BookRepository.swift]: This is a User request so you could define this in a different repository `UserRepository`

File: WBooks/Screens/BookDetails/BookDetailsView.swift [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+            rentButton.layer.cornerRadius = 20
+        }
+    }
+    @IBOutlet weak var addToWishButton: UIButton! {
+        didSet {
+            addToWishButton.setTitle("TO_WISH_LIST".localized(), for: UIControl.State.normal)
+            addToWishButton.layer.borderWidth = 2.0
+            addToWishButton.layer.cornerRadius = 20
+            addToWishButton.layer.borderColor = UIColor.strongBlue?.cgColor
+
```
Comment (carcos-wolox) [inline WBooks/Screens/BookDetails/BookDetailsView.swift]: remove line

File: WBooks/Extensions/UIStackViewExtension.swift [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+//  WBooks
+//
+//  Created by Nacho 2 on 30/04/2020.
+//  Copyright © 2020 Wolox. All rights reserved.
+//
+
+import UIKit
+
+extension UIStackView {
+    func applyCardStyle(backgroundColor: UIColor) {
```
Comment (carcos-wolox) [inline WBooks/Extensions/UIStackViewExtension.swift:12]: I think you could set this ad UIView extension 🤔

File: WBooks/Screens/BookDetails/BookDetailsViewController.swift [hunk 13]
```diff
@@ -0,0 +1,126 @@
+//
+//  BookDetailsViewController.swift
+//  WBooks
+//
+//  Created by Nacho 2 on 05/04/2020.
+//  Copyright © 2020 Wolox. All rights reserved.
+//
+
+import Foundation
```
Comment (carcos-wolox) [inline WBooks/Screens/BookDetails/BookDetailsViewController.swift]: Foundation import is not necessary if you have UIKit

File: WBooks/Screens/BookDetails/BookDetailsViewController.swift [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+        super.init(nibName: .none, bundle: .none)
+    }
+    
+    override public func loadView() {
+        view = _view
+        _view.bookTitle.text = _bookDetailsViewModel.bookModel.title
+        _view.bookGen.text = _bookDetailsViewModel.bookModel.genre
+        _view.bookYear.text = _bookDetailsViewModel.bookModel.year
+        _view.bookAuthor.text = _bookDetailsViewModel.bookModel.author
+        _view.bookStatus.text = _bookDetailsViewModel.bookModel.status
```
Comment (carcos-wolox) [inline WBooks/Screens/BookDetails/BookDetailsViewController.swift]: You could add a function in BookDetailsView to configure this

File: WBooks/Screens/BookDetails/BookDetailsView.swift [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+            commentsTable.layer.cornerRadius = 5
+            commentsTable.layer.shadowOffset = CGSize(width: 2, height: 2)
+            commentsTable.layer.shadowColor = UIColor.black.cgColor
+            commentsTable.layer.shadowOpacity = 0.2
+        }
+    }
+    
+    // MARK: - update status label
+    func updateStyles(newStatus: String) {
+        if newStatus == "Unavailable" {
```
Comment (carcos-wolox) [inline WBooks/Screens/BookDetails/BookDetailsView.swift]: I prefer an enum, but its ok

File: WBooks/Screens/BookDetails/BookDetailsViewController.swift [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+        _view.updateStyles(newStatus: _bookDetailsViewModel.bookModel.status)
+        if _bookDetailsViewModel.bookModel.image != nil {
+            _view.bookImage.loadUrl(from: _bookDetailsViewModel.bookModel.image ?? "")
+        } else {
+            _view.bookImage.image = UIImage.defaultBook
+        }
```
Comment (carcos-wolox) [inline WBooks/Screens/BookDetails/BookDetailsViewController.swift]: Same as previous PR: Set `defaultBook` always and then download the image so you always have a image

File: WBooks/Screens/BookDetails/BookDetailsViewController.swift [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+    override public func loadView() {
+        view = _view
+        _view.bookTitle.text = _bookDetailsViewModel.bookModel.title
+        _view.bookGen.text = _bookDetailsViewModel.bookModel.genre
+        _view.bookYear.text = _bookDetailsViewModel.bookModel.year
+        _view.bookAuthor.text = _bookDetailsViewModel.bookModel.author
+        _view.bookStatus.text = _bookDetailsViewModel.bookModel.status
+        _view.updateStyles(newStatus: _bookDetailsViewModel.bookModel.status)
+        if _bookDetailsViewModel.bookModel.image != nil {
+            _view.bookImage.loadUrl(from: _bookDetailsViewModel.bookModel.image ?? "")
```
Comment (carcos-wolox) [inline WBooks/Screens/BookDetails/BookDetailsViewController.swift]: In bookDetailsViewModel add a property to get this value and return nil if it's ""

Comment (carcos-wolox) [reply WBooks/Screens/BookDetails/BookDetailsViewController.swift]: Remember that the view/controller should not know anything about the model

File: WBooks/Screens/BookDetails/BookDetailsViewController.swift [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+    
+    func setupTableView() {
+        _view.commentsTable.delegate = self
+        _view.commentsTable.dataSource = self
+        _view.commentsTable.register(cell: CommentCell.self)
+        _view.commentsTable.separatorStyle = .none
+    }
+    
+    @objc private func rentButtonTapped() {
+        if _bookDetailsViewModel.bookModel.status == "Unavailable" {
```
Comment (carcos-wolox) [inline WBooks/Screens/BookDetails/BookDetailsViewController.swift]: you have the string "Unavailable" twice, so if if changes in backend you must change it everywhere, that's why I prefer the enum. Also, you must ask this to the viewModel, no to the model directly

File: WBooks/Screens/BookDetails/BookDetailsViewController.swift [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+            let alert = UIAlertController().createErrorAlert(message: "RENT_UNAVAILABLE".localized())
```
Comment (carcos-wolox) [inline WBooks/Screens/BookDetails/BookDetailsViewController.swift]: you could make createErrorAlert static, so you don't have to create an empty AlertController that you are not going to use

File: WBooks/Screens/BookDetails/BookDetailsViewController.swift [hunk 20]
```diff
@@ -0,0 +1,126 @@
...
+            _view.updateRentLoading(isLoading: true)
+            _bookDetailsViewModel.rentBook(onSuccess: rentBookSuccess, onError: rentBookError)
+        }
+    }
+}
+
+// MARK: - Extension for request callbacks
+extension BookDetailsViewController {
+    func rentBookSuccess() {
+        _view.bookStatus.text = "Unavailable"
```
Comment (carcos-wolox) [inline WBooks/Screens/BookDetails/BookDetailsViewController.swift]: you must update this from a property in viewModel. ViewModel must know it the book is available or not after rentBook request

File: WBooks/Screens/BookDetails/BookDetailsViewController.swift [hunk 21]
```diff
@@ -0,0 +1,126 @@
...
+        _view.updateStyles(newStatus: "Unavailable")
```
Comment (carcos-wolox) [inline WBooks/Screens/BookDetails/BookDetailsViewController.swift]: "Unavailable" string 👀 ~again~ chan chan chan

File: WBooks/Screens/BookDetails/BookDetailsViewController.swift [hunk 22]
```diff
@@ -0,0 +1,126 @@
...
+    
+    func rentBookError() {
+        let alert = UIAlertController().createErrorAlert(message: "RENT_ERROR".localized())
+        self.present(alert, animated: true)
+        _view.updateRentLoading(isLoading: false)
+    }
+    
+    func getCommentsSuccess() {
+        _view.updateCommentsLoading(isLoading: false)
+        if _bookDetailsViewModel.numberOfComments == 0 {
```
Comment (carcos-wolox) [inline WBooks/Screens/BookDetails/BookDetailsViewController.swift]: you could have a property in viewModel that validates this, using .isEmpty (not .count)

File: WBooks/Screens/BookDetails/BookDetailsViewModel.swift [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+//
+//  Created by Nacho 2 on 05/04/2020.
+//  Copyright © 2020 Wolox. All rights reserved.
+//
+
+import Foundation
+
+class BookDetailsViewModel {
+    let bookRepository = BookRepository()
+    let bookModel: Book
```
Comment (carcos-wolox) [inline WBooks/Screens/BookDetails/BookDetailsViewModel.swift]: When posible, this must be private

File: WBooks/Screens/BookDetails/BookDetailsViewModel.swift [hunk 24]
```diff
@@ -0,0 +1,53 @@
...
+    func getCommentCellViewModel(at indexPath: IndexPath) -> CommentCellViewModel {
+        CommentCellViewModel(commentModel: commentsList[indexPath.row])
+    }
+}
+
+extension BookDetailsViewModel {
+    func rentBook(onSuccess: @escaping () -> Void, onError: @escaping () -> Void) {
+        let rentSuccess = {
+            onSuccess()
+        }
```
Comment (carcos-wolox) [inline WBooks/Screens/BookDetails/BookDetailsViewModel.swift]: you must update your model (book status) here

File: WBooks/Screens/BookDetails/CommentCell/CommentCell.swift [hunk 25]
```diff
@@ -0,0 +1,37 @@
...
+    }
+    
+    func setCommentInformation(cellVM: CommentCellViewModel) {
+        userName.text = cellVM.username
+        userComment.text = cellVM.content
+        if let image = cellVM.image {
+            userImage.loadUrl(from: image)
+        } else {
+            userImage.image = UIImage.defaultBook
+        }
```
Comment (carcos-wolox) [inline WBooks/Screens/BookDetails/CommentCell/CommentCell.swift]: same

File: WBooks/Screens/BookDetails/BookDetailsViewModel.swift [hunk 26]
```diff
@@ -9,13 +9,13 @@
 import Foundation
 
 class BookDetailsViewModel {
-    let bookRepository = BookRepository()
-    let userRepository = UserRepository()
-    let bookModel: Book
+    private let _bookRepository = BookRepository()
+    private let _userRepository = UserRepository()
+    private var _bookModel: Book
     var commentsList: [Comment]
```
Comment (carcos-wolox) [inline WBooks/Screens/BookDetails/BookDetailsViewModel.swift]: this should be private too

File: WBooks/Screens/BookDetails/BookDetailsViewModel.swift [hunk 27]
```diff
@@ -9,13 +9,13 @@
...
     
     init(bookModel: Book) {
-        self.bookModel = bookModel
+        self._bookModel = bookModel
```
Comment (carcos-wolox) [inline WBooks/Screens/BookDetails/BookDetailsViewModel.swift]: unnecessary self

File: WBooks/Repositories/BookRepository.swift [hunk 28]
```diff
@@ -12,10 +12,12 @@ import Alamofire
 
 class BookRepository {
     
-    let url = URL(string: "https://swift-training-backend.herokuapp.com/books")!
+    let baseUrl = URL(string: "https://swift-training-backend.herokuapp.com")!
+    let header: HTTPHeaders = ["Content-Type": "application/json"]
     
-    func fetchBooks(onSuccess: @escaping ([Book]) -> Void, onError: @escaping (Error) -> Void) {
-        request(url, method: .get).responseJSON { response in
+    public func fetchBooks(onSuccess: @escaping ([Book]) -> Void, onError: @escaping (Error) -> Void) {
```
Comment (carcos-wolox) [inline WBooks/Repositories/BookRepository.swift]: unnecessary public

File: WBooks/Repositories/BookRepository.swift [hunk 29]
```diff
@@ -28,6 +30,22 @@ class BookRepository {
                 }
             }
     }
+    
+    public func fetchComments(bookId: Int, onSuccess: @escaping ([Comment]) -> Void, onError: @escaping (Error) -> Void) {
```
Comment (carcos-wolox) [inline WBooks/Repositories/BookRepository.swift]: same

File: WBooks/Repositories/UserRepository.swift [hunk 30]
```diff
@@ -0,0 +1,40 @@
...
+import Foundation
+import Result
+import Alamofire
+
+class UserRepository {
+    
+    let baseUrl = URL(string: "https://swift-training-backend.herokuapp.com")!
+    let header: HTTPHeaders = ["Content-Type": "application/json"]
+    
+    public func rentBook(rentModel: Rent, onSuccess: @escaping () -> Void, onError: @escaping (Error) -> Void) {
```
Comment (carcos-wolox) [inline WBooks/Repositories/UserRepository.swift]: same

File: WBooks/Screens/BookDetails/BookDetailsViewController.swift [hunk 31]
```diff
@@ -0,0 +1,119 @@
...
+// MARK: - Extension for request callbacks
+extension BookDetailsViewController {
+    func rentBookSuccess() {
+        _view.updateStatusStyle(newStatus: _bookDetailsViewModel.status)
+        _view.updateRentLoading(isLoading: false)
+    }
+    
+    func rentBookError() {
+        let alert = UIAlertController.createErrorAlert(message: "RENT_ERROR".localized())
+        self.present(alert, animated: true)
```
Comment (carcos-wolox) [inline WBooks/Screens/BookDetails/BookDetailsViewController.swift]: unnecessary self

File: WBooks/Screens/BookDetails/CommentCell/CommentCellViewModel.swift [hunk 32]
```diff
@@ -0,0 +1,30 @@
...
+class CommentCellViewModel {
+    let commentModel: Comment
+    
+    init(commentModel: Comment) {
+        self.commentModel = commentModel
+    }
+    
+    var image: String? {
+        guard let imageName = commentModel.image, imageName.isNotEmpty else { return nil }
+        return  imageName
```
Comment (carcos-wolox) [inline WBooks/Screens/BookDetails/CommentCell/CommentCellViewModel.swift:20]: It's better this, If you want, change BookDetailsViewModel image with this, so you don't use the `!`

**Review by carcos-wolox — APPROVED** [review]
GOOOODD!!!! 👍 💪  Just 2 small changes in comments

**Pull request merged**